### PR TITLE
WL: 21549 - Demonstrate glow outline effect on near-grabbing

### PR DESCRIPTION
**Pull request by MidnightRift**
# Info
Demonstrate glow effect on near grabbable entities. an entity that is grabbable should highlight. The effect is on by default

## Reqs
- HMD Required.

# QA

## Setup
- Install the latest Interface this PR generates
- Create Three entities one of which has the properties set to not grabbable.
- Get into HMD
## Steps - Two Person

| | Steps | Expected results |
| --- | --- | --- |
| 1. | place your controller(hand) near an Entity that is grabbable. | The entity should be highlighted. |
| 2. | place your other controller(hand) the other entity with your other hand. | Both Entities should be highlighted. |
| 3. | take your hand away from one entity | The Entity should no longer be highlighted. |
| 4. | Take your hand away from the other entity | The Entity should no longer be highlighted |
| 5. | Place both hands near an entity |  The Entity should be highlighted. |
| 6. | Remove a hand from near the entity |  The Entity should Remain highlighted. |
| 7. | Remove the other hand from near the entity |  The Entity should no longer be highlighted. |
| 8. | Place a hand near the non-grabbabale entity |  The Entity should *not* be highlighted. |

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3997/HighFidelity-Beta-PR11558-a9d41fdc16720066a2b0524cc8572f3568d475e1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3997/HighFidelity-Beta-PR11558-a9d41fdc16720066a2b0524cc8572f3568d475e1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4015/HighFidelity-Beta-PR11558-f657713ea1aaf7fc1e3ea84978b0b5e47f8eb40d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4015/HighFidelity-Beta-PR11558-f657713ea1aaf7fc1e3ea84978b0b5e47f8eb40d.exe

**Comment by PhilipRosedale** [comment]
I'm running the build but don't see any glowing for objects I am grabbing.  How do I turn the effect on?

**Comment by PhilipRosedale** [comment]
Also:  lasers and styluses do not appear for me at all in this build.

**Comment by MidnightRift** [comment]
It should just work out of the box, Ill have to take a look at why its not letting the stylus show up I tried it in Eschatology and i think i might be getting similar effects to what you mentioned, though i do not get it in my test domain. -- For now you can see what it looks like and see if that is what you wanted by trying it on the objects in hifi://Noctis/0,3,0

**Comment by Nex-Pro** [comment]
Can this be turn-offable ?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4095/HighFidelity-Beta-PR11558-f8a11ba89aff94eca3d4c621ae84816784075503.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4095/HighFidelity-Beta-PR11558-f8a11ba89aff94eca3d4c621ae84816784075503.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4116/HighFidelity-Beta-PR11558-322a7dce8a8e046aad674fe587539f16250c66d3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4116/HighFidelity-Beta-PR11558-322a7dce8a8e046aad674fe587539f16250c66d3.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4132/HighFidelity-Beta-PR11558-d0cbcd4a4837657aba015fae22c5315f94bac1be.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4132/HighFidelity-Beta-PR11558-d0cbcd4a4837657aba015fae22c5315f94bac1be.exe

**Comment by Nex-Pro** [comment]
Testing

**Comment by Nex-Pro** [comment]
### FAIL

The highlighting is fine.  steps 1 thru 8 all good. 

***BUT*** 
Unable to get lasers in this build. Several restarts, rooboted vive several times. Checked vive is ok in other builds. Without lasers it is difficult to procceed.

**Comment by Nex-Pro** [comment]
fail confirmed.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4224/HighFidelity-Beta-PR11558-772e495300a8bb4b845bc9bb6ce0fd4698a1f72e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4224/HighFidelity-Beta-PR11558-772e495300a8bb4b845bc9bb6ce0fd4698a1f72e.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4232/HighFidelity-Beta-PR11558-2808a8605edf5366bf50eaea37f972e856f6409e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4232/HighFidelity-Beta-PR11558-2808a8605edf5366bf50eaea37f972e856f6409e.exe

**Comment by Nex-Pro** [comment]
T E S T I N G win10

**Comment by Nex-Pro** [comment]
Fail on edge case ( can have two hands in two grabbable ents with only one lit ) otherwise good

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4341/HighFidelity-Beta-PR11558-052e063db297306c914b67b3b86a46b4fd1b8f19.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4341/HighFidelity-Beta-PR11558-052e063db297306c914b67b3b86a46b4fd1b8f19.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4347/HighFidelity-Beta-PR11558-d366cb99b24a0337d353802d084b5f74700e9994.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4347/HighFidelity-Beta-PR11558-d366cb99b24a0337d353802d084b5f74700e9994.exe

**Comment by Nex-Pro** [comment]
T E S T I N G  Win10

**Comment by Nex-Pro** [comment]
@MidnightRift I get no lasers in this. I did a fully clean install of this build after the problem occurred the first time. I re-ran scripts etc. Did not fix.
I then grabbed dev latest and I have lasers in that.
This is a little impractical to test as is.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4358/HighFidelity-Beta-PR11558-3a0f23b2e3572fbc2acac6e5c10e06ddc1dd8a0d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4358/HighFidelity-Beta-PR11558-3a0f23b2e3572fbc2acac6e5c10e06ddc1dd8a0d.exe

**Comment by Nex-Pro** [comment]
@MidnightRift Should I try to test this and ignore no lasers ?

**Comment by MidnightRift** [comment]
So, if i make that this (lighlightObjectsInNearGrabRange.js) lower then the priority (550) of farActionGrabEntity.js. the contorllerDispacher will still run farActionGrabEntity.js but it does nothing after the isReady state. I think this might be a bug somewhere in the controllerDispacher. I don't know specifically. -- I can leave this plugin at a higher priority for now, but it still sometimes gives the undesirable effect that Piper is talking about, I have fixed the major case where that breaks, but I can't exactly fix the timing one to my knowledge without a lower priority. There is a period of about .2 seconds after one hand leaves an entity where that edge case will still happen atm. @sethalves

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4501/HighFidelity-Beta-PR11558-4bc6f1e37f1c78ff632ecd5681ed5fbe3b90a64e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4501/HighFidelity-Beta-PR11558-4bc6f1e37f1c78ff632ecd5681ed5fbe3b90a64e.exe

**Comment by Nex-Pro** [comment]
Testing shortly by request

**Comment by Nex-Pro** [comment]
Testing Win10 and vive

**Comment by Nex-Pro** [comment]
I will leave the decision on this to the team.

Success on the test steps but there is a problem with highlighting switching off and not re-activating when the user switches to desktop then back to HMD. 

Also there is a slight edge case with overlapping entities ( I am being cruel and harsh )

Upload this JSON  
[grabOutlineTestEntities.SVO.json.zip](https://github.com/highfidelity/hifi/files/1423211/grabOutlineTestEntities.SVO.json.zip)


and put your hand in the spheres. Observe the outline flickering rapidly. Might be worth considering some cool down code.


Steps

1.	place your controller(hand) near an Entity that is grabbable.	

The entity should be highlighted. **SUCCESS**


2.	place your other controller(hand) the other entity with your other hand.	

Both Entities should be highlighted. **SUCCESS**


3.	take your hand away from one entity	 **SUCCESS**

The Entity should no longer be highlighted.


4.	Take your hand away from the other entity	

The Entity should no longer be highlighted **SUCCESS**


5.	Place both hands near an entity	The 

Entity should be highlighted. **SUCCESS**


6.	Remove a hand from near the entity	 **SUCCESS**

The Entity should Remain highlighted.


7.	Remove the other hand from near the entity	

The Entity should no longer be highlighted. **SUCCESS**


8.	Place a hand near the non-grabbabale entity	

The Entity should not be highlighted. **SUCCESS**

@midnight good work.

**Comment by MidnightRift** [comment]
The reason for the bounce is the distance to both entities are so small its hitting rounding errors, I can try calming the values.

The Outline going away after hmd to desktop, is also reproducible with marketplace items like the bow. It also happen when you ctrl + R.

File: scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        var _this = this;
+        this.hand = hand;
+        this.activeHighlightObject = null;
+
+        this.getOtherModule = function () {
+            return _this.hand ? leftHighlightNearestGrabbableEntity : rightHighlightNearestGrabbableEntity;
+        };
+
+        this.parameters = makeDispatcherModuleParameters(
+            610, // no idea what a good priority is
```
Comment (druiz17) [inline scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js]: Please remove these comments since they are not needed and if you still don't understand what these fields do, I could explain them yo you.

File: scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+
+    var leftHighlightNearestGrabbableEntity = new HighlightNearestGrabbableEntity(LEFT_HAND);
+    var rightHighlightNearestGrabbableEntity = new HighlightNearestGrabbableEntity(RIGHT_HAND);
+
+    enableDispatcherModule("leftHighlightNearestGrabbableEntity", leftHighlightNearestGrabbableEntity);
+    enableDispatcherModule("rightHighlightNearestGrabbableEntity", rightHighlightNearestGrabbableEntity);
+
+    this.cleanup = function () {
```
Comment (druiz17) [inline scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js]: this.clean will pollute the global namespace (other modules break this rule).  Instead do  ```function clean()``` to prevent from other scripts from overriding ```this.clean``` variable

File: scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                    }
+
+                    if (userData && userData.grabbableKey) {
+                        isGrabbable = (userData.grabbableKey.grabbable);
+                        isCloneable = (userData.grabbableKey.cloneable);
+                    }
+                }
+            }
+
+            if (nearbyEntity && (isGrabbable || isCloneable) && _this.activeHighlightObject === null &&
```
Comment (druiz17) [inline scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js]: These If statement are a bit difficult to read. I think you could simplify code.

File: scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                (_this.getOtherModule().activeHighlightObject === null ||
+                    (_this.getOtherModule().activeHighlightObject &&
+                        _this.activeHighlightObject.id !== _this.getOtherModule().activeHighlightObject.id))) {
+
+                Selection.removeFromSelectedItemsList("contextOverlayHighlightList", 'entity', _this.activeHighlightObject.id);
+                _this.activeHighlightObject = null;
+
+            }
+
+            return makeRunningValues(false, [], []); // not sure why but this needs to be false to let the stylus and hands and stuff work
```
Comment (druiz17) [inline scripts/system/controllers/controllerModules/highlightObjectInNearGrabRange.js]: Please remove these comments since they are not needed and if you still don't understand what these fields do, I could explain them yo you.

**Comment by themelissabrown** [comment]
Once this is completed, PM's should test and approve it prior to merging

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4659/HighFidelity-Beta-PR11558-b0582c0fdb84e429b24be4ed6f4c7f4172e9cac4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4659/HighFidelity-Beta-PR11558-b0582c0fdb84e429b24be4ed6f4c7f4172e9cac4.exe

**Comment by PhilipRosedale** [comment]
Hi gang - sorry I've been busy and still haven't had a chance to review this.  Any chance you could bring it up to current with latest version?

**Comment by PhilipRosedale** [comment]
I still can't get the effect to work after starting and switching into HMD

**Comment by PhilipRosedale** [comment]
Actually it seems that some models don't show the effect and I am not sure why.  Balloons from marketplace have no highlight, and the dice on the gaming table don't although the chess pieces and checker pieces do have highlights.

**Comment by MidnightRift** [comment]
I checked the balloon and its being brought into the world with missing user data. This Highlighting code is checking to see if an object is grabbable via `grabableKey.grabbable = true` .

**Comment by themelissabrown** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6162/app_11558_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6162/HighFidelity-Beta-PR11558-b0582c0fdb84e429b24be4ed6f4c7f4172e9cac4.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6162/HighFidelity-Beta-PR11558-b0582c0fdb84e429b24be4ed6f4c7f4172e9cac4.dmg

**Comment by mukulHF** [comment]
I tried it in Sandbox with small spheres lying on the floor, they are not highlighting. I checked in create> properties and they are grabbable. @MidnightRift can you plz take a look at it?

**Comment by themelissabrown** [comment]
Closing until we can figure out how to utilize.
